### PR TITLE
[FSDP] fix: fix for fsdp zero2 validation error

### PR DESCRIPTION
**Pull request by Edwiv**
# Problem
When sharding_strategy is set to SHARD_GRAD_OP and forward_prefetch is turned on, the validation after the train has an incorrect weight shape.
image

# Analyze
When using `SHARD_GRAD_OP`, the `free_unsharded_flat_param` in `_post_forward_reshard` is often False, so it does not set the handle's `_prefetched` flag to False after the forward.

The normal train phase sets this flag to False in the `_post_backward_final_callback`, and the validation phase doesn't execute the hook, so after the first iter of the validation is done, the flag of the handle of the prefetched will remain True.

This will cause the handle to skip the `_unshard` in the next `_pre_forward_unshard`, and the `_prefetch_handle` will not do a prefetch, which will result in an incorrect weight shape.

### CODE REVIEW

**Comment by Edwiv** [comment]
@awgu sure~

**Comment by awgu** [comment]
@Edwiv feel free to put it in test_fsdp_misc.py:
https://github.com/pytorch/pytorch/blob/fe11227764dbfc384b2aa23c8f0aa00fb8038bdc/test/distributed/fsdp/test_fsdp_misc.py#L71

(I do not think we have a place for eval tests right now.)

**Comment by Edwiv** [comment]
@awgu done

**Comment by awgu** [comment]
@Edwiv Thanks for adding the unit test! It will run automatically in CI.

Would it be possible to strengthen the unit test to check the correctness, e.g. with DDP?

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 1]
```diff
@@ -143,6 +143,58 @@ def _check_device_matches(module, device_id):
...
+        )
+        x = torch.randn(8, 1, 28, 28, device='cuda')
+        y = torch.randint(low=0,high=9,size=(8,), device='cuda')
+
+        for _ in range(5):
+            loss = model(x, y)
+            loss = loss.sum()
+            loss.backward()
+
+        with torch.no_grad():
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_misc.py:193]: are we interested in putting the model back to train after eval, and verify that training still works fine?

**Review by rohan-varma — COMMENTED** [review]
thanks for the fix!

File: test/distributed/fsdp/test_fsdp_misc.py (hunk 1 above)
Comment (Edwiv) [reply test/distributed/fsdp/test_fsdp_misc.py:193]: done~

**Comment by awgu** [comment]
> Would it be possible to strengthen the unit test to check the correctness, e.g. with DDP?

Sorry, I wanted to bump this in case you missed it @Edwiv!

**Comment by Edwiv** [comment]
@awgu Sorry I didn't understand how to test with DDP, what should be done to strengthen the unit testing?

**Comment by awgu** [comment]
@Edwiv You can construct an identical model with DDP applied (to implement data parallel semantics). Then, you can the same training loop for both DDP and FSDP models and compare each iteration's loss (for example) to check for correctness.

**Comment by Edwiv** [comment]
@awgu I got it, I've added the DDP comparison to the unit test, please take a look.

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 2]
```diff
@@ -143,6 +143,101 @@ def _check_device_matches(module, device_id):
...
+            auto_wrap_policy=ModuleWrapPolicy([nn.Linear, nn.Conv2d]),
+        )
+        ddp_model = torch.nn.parallel.DistributedDataParallel(
+            model1,
+        )
+        x = torch.randn(8, 1, 28, 28, device="cuda").requires_grad_()
+        y = torch.randint(low=0, high=9, size=(8,), device="cuda")
+        x1 = x.clone().detach().requires_grad_()
+        y1 = y.clone().detach()
+        seed = 20231010
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:195]: We should set the seed to be something like `self.rank + 1` (anything so that data parallel ranks have different seeds); otherwise, the gradient reduction may not be tested (something like `x + x / 2` = `x` is the same as not summing and dividing).

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 3]
```diff
@@ -143,6 +143,101 @@ def _check_device_matches(module, device_id):
...
+
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed(seed)
+        for _ in range(5):
```
[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_misc.py:199]: nit: `fsdp_loss` and `ddp_loss` might be more informative names than `loss` and `loss1`.

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 4]
```diff
@@ -143,6 +143,101 @@ def _check_device_matches(module, device_id):
...
+            loss1 = ddp_model(x1, y1)
+            loss1 = loss1.sum()
+            loss1.backward()
+
+        assert torch.allclose(loss, loss1)
+        assert torch.allclose(x.grad.data, x1.grad.data)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:212]: nit: the `.data` are not needed.

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 5]
```diff
@@ -143,6 +143,109 @@ def _check_device_matches(module, device_id):
...
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed(seed)
+        for _ in range(5):
+            ddp_loss = ddp_model(x1, y1)
+            ddp_loss = ddp_loss.sum()
+            ddp_loss.backward()
+            ddp_opt.step()
+
+        assert torch.allclose(fsdp_loss, ddp_loss)
+        assert torch.allclose(x.grad, x1.grad)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:219]: nit: Would it be possible to check correctness after every iteration? This lets us narrow down / catch bugs much faster.

You can reference: https://github.com/pytorch/pytorch/pull/110948/files#diff-ab5af580410c642dd66ea27656265fbbc1ec6c6713e048a0ef111573eb52286cR195

Comment (Edwiv) [reply test/distributed/fsdp/test_fsdp_misc.py]: I've perfected this UT, and by the way, the author of the PR above turned out to be a coworker of mine😂.

**Review by awgu — APPROVED** [review]

**Comment by awgu** [comment]
Lint error looks like an infra issue. Let me rebase and merge.
